Add import processes

inv-jishnu opened the pull request:
## Description

In this PR I have added import processes to process the import file based on the file format and related dtos and util files for it.

## Related issues and/or PRs
 Please review this PR once the below PRs are reviewed and merged and master branch is merged to this branch with those changes.
 - https://github.com/scalar-labs/scalardb/pull/2442
 - https://github.com/scalar-labs/scalardb/pull/2446


## Changes made

Adde classes to process the import source file based on the file format and related dtos and util classes

## Checklist

> The following is a best-effort checklist. If any items in this checklist are not applicable to this PR or are dependent on other, unmerged PRs, please still mark the checkboxes after you have read and understood each item.

- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation to reflect the changes.
- [ ] Any remaining open issues linked to this PR are documented and up-to-date (Jira, GitHub, etc.).
- [ ] Tests (unit, integration, etc.) have been added for the changes.
- [ ] My changes generate no new warnings.
- [ ] Any dependent changes in other PRs have been merged and published.

## Additional notes (optional)

Road map to merge remaining data loader core files. Current status

* **General** 
  * Add ScalarDB Dao and related files - https://github.com/scalar-labs/scalardb/pull/2417
  * TableMetadataService(partially replaced by ConsensusUtils): https://github.com/scalar-labs/scalardb/pull/2434

* **Export**
  * Export options and validations: https://github.com/scalar-labs/scalardb/pull/2435
  * ProducerTasks: 1 PR incoming
  
* **Import**
  *  Dto classes and utilities: https://github.com/scalar-labs/scalardb/pull/2442
  * Import processor and task code:  2-3 PRs incoming
    * https://github.com/scalar-labs/scalardb/pull/2446
  * Code for Import transaction batch and  data chunk import:  1 PR Incoming 
  * ControlFile related Dtos: https://github.com/scalar-labs/scalardb/pull/2445
  * Import logger: 1 PR incoming
  * Import Manager:  1 PR incoming

## Release notes

NA

komamitsu commented on data-loader/core/src/main/java/com/scalar/db/dataloader/core/dataimport/task/ImportTask.java:
```diff
@@ -0,0 +1,431 @@
+          .importAction(importAction)
+          .status(ImportTargetResultStatus.DATA_ALREADY_EXISTS)
+          .errors(
+              Collections.singletonList(CoreError.DATA_LOADER_DATA_ALREADY_EXISTS.buildMessage()))
```
How about adding key information in the error message for investigation?

komamitsu replied:
Thanks for the explanation! Sounds good.

How about adding a comment at https://github.com/scalar-labs/scalardb/pull/2462/files#diff-fd06988f4ced6a0d6d0b1049c69ab07df11c372c3ccb868a78afbe4d24e6f663R827 like `// These error messages will be output with key information` ?

komamitsu commented on data-loader/core/src/main/java/com/scalar/db/dataloader/core/dataimport/task/ImportTask.java:
```diff
@@ -0,0 +1,431 @@
+          .importedRecord(mutableSourceRecord)
+          .importAction(importAction)
+          .status(ImportTargetResultStatus.DATA_NOT_FOUND)
+          .errors(Collections.singletonList(CoreError.DATA_LOADER_DATA_NOT_FOUND.buildMessage()))
```
Same as above

komamitsu commented on data-loader/core/src/main/java/com/scalar/db/dataloader/core/dataimport/task/ImportTask.java:
```diff
@@ -0,0 +1,431 @@
+      return ImportTargetResult.builder()
+          .namespace(namespace)
+          .tableName(table)
+          .status(ImportTargetResultStatus.VALIDATION_FAILED)
```
Probably same as above? Could you take a look at those error messages added by this PR from the view point of users who need to investigate failures?

komamitsu commented on data-loader/core/src/main/java/com/scalar/db/dataloader/core/util/ColumnUtils.java:
```diff
@@ -112,4 +122,128 @@ public static Column<?> createColumnFromValue(
+   * @throws ColumnParsingException if an error occurs while base64 parsing the column
+   */
+  private static Column<?> getColumn(
+      Result scalarDBResult,
```
```suggestion
      @Nullable Result scalarDBResult,
```

komamitsu commented on data-loader/core/src/main/java/com/scalar/db/dataloader/core/dataimport/processor/ImportProcessor.java:
```diff
@@ -0,0 +1,419 @@
+import org.slf4j.LoggerFactory;
+
+@RequiredArgsConstructor
+public abstract class ImportProcessor {
```
ImportProcessor and its child classes are a bit complicated. They don't need to have unit tests?

inv-jishnu commented on data-loader/core/src/main/java/com/scalar/db/dataloader/core/dataimport/processor/CsvImportProcessor.java:
```diff
@@ -0,0 +1,175 @@
+                  } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                    throw new RuntimeException(
+                        CoreError.DATA_LOADER_DATA_CHUNK_PROCESS_FAILED.buildMessage(
```
Currently here there is InterruptedException may only be thrown by `dataChunkQueue.poll`.
Please do let me know if you feel there is need to handle any other exception here.
Thank you

komamitsu replied:
I see. That sounds reasonable basically since `processDataChunk` is called asynchronously. (OOMException could still be thrown, though)

inv-jishnu commented:
@komamitsu san,

I have added changes as suggested including adding some basic unit test cases apart from the feedback related to error message clarity (I have added some more details for that as a comment).
Please take a look at this again and please do let me know if any further changes are required.
Thank you.

brfrn169 commented on data-loader/core/src/main/java/com/scalar/db/dataloader/core/dataimport/ImportManager.java:
```diff
@@ -0,0 +1,187 @@
+  @NonNull private final ImportOptions importOptions;
+  private final ImportProcessorFactory importProcessorFactory;
+  private final List<ImportEventListener> listeners = new ArrayList<>();
+  private final ScalarDBMode scalarDBMode;
```
It looks like I missed this in the previous PR review, but we should rename this to `ScalarDbMode`. We should also rename `ScalarDBDao` and `ScalarDBDaoException` accordingly.

Could you please handle this in a separate PR?

inv-jishnu replied:
I will create another PR for the changes mentioned.
Thank you.

inv-jishnu replied:
@brfrn169 san,
I have created PR https://github.com/scalar-labs/scalardb/pull/2582 to address this. I will mark it ready for review once this PR is merged.
Thank you.

brfrn169 commented on data-loader/core/src/main/java/com/scalar/db/dataloader/core/dataimport/processor/ImportProcessor.java:
```diff
@@ -0,0 +1,457 @@
+    Instant startTime = Instant.now();
+    List<ImportTransactionBatch> transactionBatches =
+        splitIntoTransactionBatches(dataChunk, transactionBatchSize);
+    ExecutorService transactionBatchExecutor =
+        Executors.newFixedThreadPool(Math.min(transactionBatches.size(), numCores * 2));
```
Could you please explain why we use `Math.min(transactionBatches.size(), numCores * 2)` here?

inv-jishnu replied:
@brfrn169 san,
I added this so that the number of threads created is not too much. The initial way was `numCores * 2`, I updated this considering if the transaction size provided may be small then we can create less threads. Please do let me know if you think any changes are needed here.
Thank you.

komamitsu replied:
Determining concurrency based on the number of cores is useful for CPU-bound operations, but since the import tool is I/O-bound, that may not be the best approach. Also, appropriate concurrency depends on the underlying database's latency and stability, and the number of transaction batches. Probably, it would be great if it's configurable.

inv-jishnu replied:
@komamitsu san,
I will add the changes to make it configurable.
Thank you.

inv-jishnu replied:
@komamitsu san,
I have added changes to make it configurable.
Thank you.

brfrn169 reviewed (commented):
Left several comments, but I haven’t reviewed all the code yet. I’ll continue reviewing this PR. Thanks.

komamitsu commented on data-loader/core/src/main/java/com/scalar/db/dataloader/core/dataimport/processor/JsonImportProcessor.java:
```diff
@@ -45,6 +44,7 @@ public class JsonImportProcessor extends ImportProcessor {
 
   private static final DataLoaderObjectMapper OBJECT_MAPPER = new DataLoaderObjectMapper();
   private static final AtomicInteger dataChunkIdCounter = new AtomicInteger(0);
+  private static final int MAX_QUEUE_SIZE = 10;
```
This queue size might be too less when using a large number of cores like 32 cores. I guess this limitation is to prevent OOM. If so, 256 or 512 should be okay. Maybe it would be great if this is configurable.

I think you can decide whether to change this after conducting some performance tests with various numbers of cores.

inv-jishnu replied:
@komamitsu san,
Thank you for the suggestion, I have now set it to 256. We are working on performance tests once that is completed I will update this accordingly.

inv-jishnu replied:
@komamitsu san,
I have made this value configurable. I will set a suitable value once the performance testing is completed. For now it is set to `256`.

komamitsu commented on data-loader/core/src/test/java/com/scalar/db/dataloader/core/dataimport/processor/CsvImportProcessorTest.java:
```diff
@@ -0,0 +1,119 @@
+import org.mockito.Mock;
+import org.mockito.Mockito;
+
+public class CsvImportProcessorTest {
```
[minor] This can be package-private.

komamitsu commented on data-loader/core/src/test/java/com/scalar/db/dataloader/core/dataimport/processor/CsvImportProcessorTest.java:
```diff
@@ -0,0 +1,119 @@
+import org.mockito.Mockito;
+
+public class CsvImportProcessorTest {
+  @Mock ImportProcessorParams params;
```
[minor] This can be `private`

inv-jishnu commented on data-loader/core/src/main/java/com/scalar/db/dataloader/core/dataimport/processor/CsvImportProcessor.java:
```diff
@@ -0,0 +1,209 @@
+  public ConcurrentHashMap<Integer, ImportDataChunkStatus> process(
+      int dataChunkSize, int transactionBatchSize, BufferedReader reader) {
+    int numCores = Runtime.getRuntime().availableProcessors();
+    ExecutorService dataChunkExecutor = Executors.newFixedThreadPool(numCores);
```
@brfrn169 san,
When I checked I found that newSingleThreadExecutor() uses an unbounded queue, which may lead to out-of-memory errors when tthere are too many tasks. Using `Executors.newFixedThreadPool` limits the number of queued tasks and reduces the chance of this. Also I saw some other drawbacks such as newSingleThreadExecutor executes tasks sequentially.I also saw Executors.newSingleThreadExecutor() is better to handle thread safe resources. The ` Executors.newFixedThreadPool` lets us execute tasks parallel and since tasks don't depend on each other I thought  Executors.newFixedThreadPool is better to be used in this case. 

Please do let me know if this needs to be updated to `Executors.newSingleThreadExecutor()` or if you see some improvements that can be added here.
Thank you.

brfrn169 replied:
> Using Executors.newFixedThreadPool limits the number of queued tasks and reduces the chance of this.

No, `Executors.newFixedThreadPool` also uses an unbounded queue.

> Also I saw some other drawbacks such as newSingleThreadExecutor executes tasks sequentially

I meant that this `dataChunkExecutor` seems to be used only in the following place:

https://github.com/scalar-labs/scalardb/pull/2462/files#diff-2c35c36faf50b3854da17eaa09fbea7dd6fa8c116f8b9872bc0e97c6cdd4a83eR75-R77

So, we’re effectively using only one thread of this `dataChunkExecutor`, right? That's why I think we can use `newSingleThreadExecutor()` here. 

What do you think?

inv-jishnu replied:
@brfrn169 san,
Thank you. Understood. Sorry, I mixed up `Executors.newFixedThreadPool` (used in the beginning) and `Executors.newSingleThreadExecutor`. What you said makes sense. I have updated it.

komamitsu commented on data-loader/core/src/main/resources/config.properties:
```diff
@@ -0,0 +1,2 @@
+transaction.batch.thread.pool.size=16
+import.data.chunk.queue.size=256
```
[minor] A file should end with a newline https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_206

inv-jishnu replied:
I have added a new line.
Thank you.

komamitsu reviewed (approved):
Left a minor comment. But other than that, LGTM! 👍

brfrn169 commented on data-loader/core/src/main/java/com/scalar/db/dataloader/core/util/ConfigUtil.java:
```diff
@@ -0,0 +1,98 @@
+ * <p>This class reads properties from a {@code config.properties} file located in the classpath.
+ */
+public class ConfigUtil {
+  public static final String CONFIG_PROPERTIES = "config.properties";
```
So, are we using a fixed configuration file? Could you please explain why we don’t use command-line arguments for the queue size and thread pool size?

Anyway, if we’re using a file for the configurations, I think we should pass the configuration file name via command-line arguments.

komamitsu replied:
> I think we should pass the configuration file name via command-line arguments.

👍

inv-jishnu replied:
@brfrn169 san, @komamitsu san,
Thank you for the suggestions.
I have updated to use command line arguments to configure queue size(added new parameter) and thread pool size (there was an existing parameter for this, but I didn't use that as the initial new data loader changes had usage of number of cores in it. So I didn't change that). I have removed configurable by properties file completely. Asking the users to add a properties file for just 2 parameter while rest are passed as arguments seemed confusing. I have also remove `numCores` from method arguments and instead use the value directly as instructed.

inv-jishnu commented:
@brfrn169 san,
I have made further changes based on feedback.
Please take a look at this again when you get a chance.
Thank you.

brfrn169 reviewed (approved):
LGTM! Thank you!

ypeckstadt reviewed (approved):
LGTM. Thank you.

Torch3333 reviewed (approved):
LGTM, thank you!
